nmt.go

```diff
@@ -142,71 +149,119 @@ func (n *NamespacedMerkleTree) ProveRange(start, end int) (Proof, error) {
+	// TODO [Me] however, we may want to actually add an extra field to the `Proof`
+	// TODO [Me] that indicates whether the proof is for inclusion or for absence
+
 	return NewAbsenceProof(proofStart, proofEnd, proof, n.leafHashes[proofStart], isMaxNsIgnored), nil
 }
 
+// buildRangeProof returns the nodes (as byte slices) in the range proof of the supplied range i.e.,
+// [proofStart, proofEnd) where proofEnd is non-inclusive.
+// The nodes are ordered according to in order traversal of the namespaced tree. (TODO [Me] I need to double-check the order)
 func (n *NamespacedMerkleTree) buildRangeProof(proofStart, proofEnd int) [][]byte {
-	proof := [][]byte{}
+	// TODO a more secure way would be to make it slice of fixed size arrays i.e., the hash output size
+	proof := [][]byte{} // it is the list of nodes hashes (as byte slices) with no index
 	var recurse func(start, end int, includeNode bool) []byte
+
+	// start, end are indices of leaves in the tree hence they should be within the size of the tree i.e.,
+	// less than or equal to the len(n.leaves)
+	// this recursive algorithm is inspired by RFC6962 https://www.rfc-editor.org/rfc/rfc6962#section-2.1
+	// includeNode indicates whether the hash of the current subtree (covering the supplied range i.e., [start, end)) or
+	// one of its constituent subtrees should be part of the proof
 	recurse = func(start, end int, includeNode bool) []byte {
-		if start >= len(n.leafHashes) {
+		if start >= len(n.leafHashes) { // TODO [Me] why against leafHashes? and not leaves?
+			// TODO [Me] Why there is no check for the end index?
 			return nil
 		}
 
 		// reached a leaf
 		if end-start == 1 {
 			leafHash := n.leafHashes[start]
-			// if current range does not overlap with proof range, add a node to proofs
+			// if the index of the leaf node is out of the queried range i.e., [proofStart, proofEnd]
+			// and if the leaf is required as part of the proof i.e., includeNode == true
 			if (start < proofStart || start >= proofEnd) && includeNode {
+				// add the leafHash to the proof
 				proof = append(proof, leafHash)
 			}
+			// if the index of the leaf is within the queried range i.e., [proofStart, proofEnd] OR
+			// if the leaf is not required as part of the proof i.e., includeNode == false
 			return leafHash
 		}
 
-		// recursively get left and right subtree
+
+		// newIncludeNode indicates whether one of the subtrees of the current subtree [start, end)
+		// may have an overlap with the queried proof range i.e., [proofStart, proofEnd)
 		newIncludeNode := includeNode
-		if (end <= proofStart || start >= proofEnd) && includeNode {
+		// check whether the subtree representing the [start, end) range of leaves has overlap with the
+		// queried proof range i.e., [proofStart, proofEnd)
+		// if not
+		if (end <= proofStart || start >= proofEnd) && includeNode { //TODO [Me] the `&& includeNode` seems ineffective and unnecessary
```

Comment:
Not sure I understand this 🤔

Reply:
My suggestion is about a small simplification on the if statement.
What I meant is that  if we remove the condition on the `includeNode` from the `if` statement, we could get the same result, i.e.,
```
newIncludeNode := includeNode
if (end <= proofStart || start >= proofEnd) {
			newIncludeNode = false
		}
``` 
The reason is that:
In the original version, If `includeNode` is `false`,  the `if` block is never reached, hence the `newIncludeNode` contains the `includeNode` value i.e., `false` (the same result is yield in the new if statement)
In the original version, if the `includeNode` is `true`, and if the first condition is `true`  i.e., `(end <= proofStart || start >= proofEnd)` then we enter the if block and toggle the  `newIncludeNode` to `false` (which is the same result we obtain from the new if statement).